File: apps/ios/Anky/Core/Mirror/MirrorClient.swift
import Foundation
#if SWIFT_PACKAGE
import AnkyProtocol
#endif

struct MirrorClient {
    let baseURL: URL
    var session: URLSession = .shared

    enum Intent: String {
        case reflection
        case nudge
    }

    func askAnky(
        bytes: Data,
        identity: WriterIdentity,
        appVersion: String? = nil,
        intent: Intent = .reflection,
        surface: String? = nil,
        progress: ((MirrorProgressEvent) async -> Void)? = nil,
        reflectionChunk: ((MirrorReflectionChunkEvent) async -> Void)? = nil
    ) async throws -> MirrorResponsePayload {
        let request = try makeRequest(
            bytes: bytes,
            identity: identity,
            appVersion: appVersion,
            intent: intent,
            surface: surface
        )
        let (stream, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw MirrorClientError.invalidResponse
        }

        if !(200..<300).contains(http.statusCode) {
            let data = try await Self.collect(stream)
            if let envelope = try? JSONDecoder().decode(MirrorErrorEnvelope.self, from: data) {
                throw MirrorClientError.server(envelope.error.payload)
            }
            throw MirrorClientError.server(.fallback)
        }

        var currentEvent: String?
        var dataLines: [String] = []

        func flushEvent() async throws -> MirrorResponsePayload? {
            guard let currentEvent else {
                dataLines.removeAll()
                return nil
            }
            let payload = dataLines.joined(separator: "\n")
            dataLines.removeAll()
            switch currentEvent {
            case "update":
                if let data = payload.data(using: .utf8),
                   let event = try? JSONDecoder().decode(MirrorProgressEvent.self, from: data) {
                    await progress?(event)
                }
                return nil
            case "reflection_chunk":
                if let data = payload.data(using: .utf8),
                   let event = try? JSONDecoder().decode(MirrorReflectionChunkEvent.self, from: data) {
                    await reflectionChunk?(event)
                }
                return nil
            case "reflection":
                guard let data = payload.data(using: .utf8),
                      let event = try? JSONDecoder().decode(MirrorReflectionEvent.self, from: data) else {
                    throw MirrorClientError.invalidResponse
                }
                let reflection = event.markdown.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !reflection.isEmpty else {
                    throw MirrorClientError.invalidResponse
                }
                let hash = event.headers.value(forHTTPHeaderField: "X-Anky-Hash") ?? AnkyHasher.sha256Hex(bytes)
                let tags = event.tags ?? event.headers
                    .value(forHTTPHeaderField: "X-Anky-Tags")
                    .flatMap(Self.tags)
                    ?? []
                return MirrorResponsePayload(
                    hash: hash,
                    title: Self.title(fromMarkdown: reflection),
                    reflection: reflection,
                    tags: tags
                )
            case "error":
                throw MirrorClientError.server(Self.errorPayload(fromSSEPayload: payload))
            default:
                return nil
            }
        }

        for try await line in stream.lines {
            let trimmedLine = line.trimmingCharacters(in: .whitespaces)
            if trimmedLine.isEmpty {
                if let payload = try await flushEvent() {
                    return payload
                }
                currentEvent = nil
                continue
            }
            if trimmedLine.hasPrefix("event:") {
                if currentEvent != nil {
                    if let payload = try await flushEvent() {
                        return payload
                    }
                }
                currentEvent = String(trimmedLine.dropFirst("event:".count)).trimmingCharacters(in: .whitespaces)
            } else if trimmedLine.hasPrefix("data:") {
                dataLines.append(String(trimmedLine.dropFirst("data:".count)).trimmingCharacters(in: .whitespaces))
            }
        }

        if let payload = try await flushEvent() {
            return payload
        }
        throw MirrorClientError.invalidResponse
    }

    private func makeRequest(
        bytes: Data,
        identity: WriterIdentity,
        appVersion: String?,
        intent: Intent,
        surface: String? = nil
    ) throws -> URLRequest {
        let signed = try AnkyPostSigner.sign(body: bytes, identity: identity)
        var request = URLRequest(url: baseURL.appendingPathComponent("anky"))
        request.httpMethod = "POST"
        request.httpBody = bytes
        request.setValue("text/plain; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue(signed.identityVersion, forHTTPHeaderField: "X-Anky-Identity-Version")
        request.setValue(signed.accountId, forHTTPHeaderField: "X-Anky-Account")
        request.setValue(signed.signatureType, forHTTPHeaderField: "X-Anky-Signature-Type")
        request.setValue(signed.signature, forHTTPHeaderField: "X-Anky-Signature")
        request.setValue(signed.requestTime, forHTTPHeaderField: "X-Anky-Request-Time")
        request.setValue(signed.client, forHTTPHeaderField: "X-Anky-Client")
        request.setValue(intent.rawValue, forHTTPHeaderField: "X-Anky-Intent")
        if let appVersion {
            request.setValue(appVersion, forHTTPHeaderField: "X-Anky-App-Version")
        }
        if let surface {
            request.setValue(surface, forHTTPHeaderField: "X-Anky-Surface")
        }
        return request
    }

    private static func collect(_ bytes: URLSession.AsyncBytes) async throws -> Data {
        var data = Data()
        for try await byte in bytes {
            data.append(contentsOf: [byte])
        }
        return data
    }

    private static func tags(_ value: String) -> [String]? {
        guard let data = value.data(using: .utf8),
              let tags = try? JSONDecoder().decode([String].self, from: data) else {
            return nil
        }
        return tags
    }

    private static func errorPayload(fromSSEPayload payload: String) -> MirrorServerErrorPayload {
        guard let data = payload.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return .fallback
        }
        if let body = object["body"] as? [String: Any],
           let error = body["error"] as? [String: Any],
           let message = error["message"] as? String {
            return MirrorServerErrorPayload(code: error["code"] as? String, message: message)
        }
        if let message = object["message"] as? String {
            return MirrorServerErrorPayload(code: object["code"] as? String, message: message)
        }
        return .fallback
    }

    private static func title(fromMarkdown markdown: String) -> String {
        let lines = markdown.split(whereSeparator: \.isNewline).map(String.init)
        let heading = lines.first { $0.hasPrefix("# ") }?
            .dropFirst(2)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let fallback = lines.first?.trimmingCharacters(in: .whitespacesAndNewlines)
        let title = heading?.isEmpty == false ? heading : fallback
        return title?.isEmpty == false ? title! : "reflection"
    }
}

struct MirrorProgressEvent: Codable, Equatable {
    let stage: String
    let message: String?
}

struct MirrorReflectionChunkEvent: Codable, Equatable {
    let chunk: String
    let generatedCharacters: Int
}

struct MirrorResponsePayload: Codable, Equatable {
    let hash: String
    let title: String
    let reflection: String
    let tags: [String]
}

struct MirrorServerErrorPayload: Equatable {
    let code: String?
    let message: String

    static let fallback = MirrorServerErrorPayload(
        code: nil,
        message: "Anky could not return a reflection right now."
    )

    var isEntitlementDenied: Bool {
        // ENTITLEMENT_REQUIRED is the boundary's only denial for a
        // non-entitled account. Free clients never ask, so this is the
        // defensive mapping for stale state: it reads as the veil, never
        // as an error.
        code == "ENTITLEMENT_REQUIRED"
    }
}

enum MirrorClientError: Error, LocalizedError, Equatable {
    case invalidURL
    case invalidResponse
    case hashMismatch
    case server(MirrorServerErrorPayload)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "The mirror URL is not valid."
        case .invalidResponse:
            return "The mirror returned an invalid response."
        case .hashMismatch:
            return "The mirror response did not match this .anky."
        case .server(let payload):
            return payload.message
        }
    }

    var serverPayload: MirrorServerErrorPayload? {
        guard case .server(let payload) = self else { return nil }
        return payload
    }
}

private struct MirrorErrorEnvelope: Decodable {
    let error: MirrorError
}

private struct MirrorError: Decodable {
    let code: String?
    let message: String

    var payload: MirrorServerErrorPayload {
        MirrorServerErrorPayload(code: code, message: message)
    }
}

private struct MirrorReflectionEvent: Decodable {
    let markdown: String
    let tags: [String]?
    let headers: [String: String]
}

private extension Dictionary where Key == String, Value == String {
    func value(forHTTPHeaderField field: String) -> String? {
        first { key, _ in key.caseInsensitiveCompare(field) == .orderedSame }?.value
    }
}

File: apps/write-8-minutes/android/app/src/main/java/inc/anky/android/core/mirror/MirrorClient.kt
package inc.anky.android.core.mirror

import inc.anky.android.core.identity.AnkyPostSigner
import inc.anky.android.core.identity.WriterIdentity
import inc.anky.android.core.protocol.AnkyHasher
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

class MirrorClient(
    private val configuration: MirrorConfiguration,
    private val client: OkHttpClient = OkHttpClient(),
) {
    fun askAnky(
        bytes: ByteArray,
        identity: WriterIdentity,
        appVersion: String? = null,
        intent: MirrorIntent = MirrorIntent.Reflection,
        progress: ((MirrorProgressEvent) -> Unit)? = null,
        reflectionChunk: ((MirrorReflectionChunkEvent) -> Unit)? = null,
    ): MirrorResponsePayload {
        val signed = AnkyPostSigner.sign(body = bytes, identity = identity)
        val base = configuration.effectiveBaseUrl().trimEnd('/')
        if (!base.startsWith("http://") && !base.startsWith("https://")) {
            throw MirrorClientError.InvalidUrl
        }
        val url = "$base/anky".toHttpUrlOrNull() ?: throw MirrorClientError.InvalidUrl
        val request = Request.Builder()
            .url(url)
            .post(bytes.toRequestBody("text/plain; charset=utf-8".toMediaType()))
            .header("Content-Type", "text/plain; charset=utf-8")
            .header("Accept", "text/event-stream")
            .header("X-Anky-Identity-Version", signed.identityVersion)
            .header("X-Anky-Account", signed.accountId)
            .header("X-Anky-Signature-Type", signed.signatureType)
            .header("X-Anky-Signature", signed.signature)
            .header("X-Anky-Request-Time", signed.requestTime)
            .header("X-Anky-Client", signed.client)
            .header("X-Anky-Intent", intent.headerValue)
            .apply {
                if (!appVersion.isNullOrBlank()) {
                    header("X-Anky-App-Version", appVersion)
                }
            }
            .build()

        client.newCall(request).execute().use { response ->
            val body = response.body ?: throw MirrorClientError.InvalidResponse
            val bodyText = if (response.isSuccessful && response.isEventStream()) null else body.string()
            if (response.isSuccessful) {
                val payload = if (response.isEventStream()) {
                    parseEventStream(
                        source = body.source(),
                        bytes = bytes,
                        progress = progress,
                        reflectionChunk = reflectionChunk,
                    )
                } else {
                    val reflection = bodyText?.trim()?.takeIf { it.isNotEmpty() }
                        ?: throw MirrorClientError.InvalidResponse
                    MirrorResponsePayload(
                        hash = response.header("X-Anky-Hash") ?: AnkyHasher.sha256Hex(bytes),
                        title = titleFromMarkdown(reflection),
                        reflection = reflection,
                        tags = response.header("X-Anky-Tags").tags().orEmpty(),
                        creditsRemaining = response.header("X-Anky-Credits-Remaining").creditsRemaining(),
                    )
                }
                if (payload.hash != AnkyHasher.sha256Hex(bytes)) throw MirrorClientError.HashMismatch
                return payload
            }
            throw bodyText
                ?.toJsonObjectOrNull()
                ?.toMirrorServerError()
                ?: MirrorClientError.Server(
                    MirrorErrorCode.Unknown,
                    "Anky could not return a reflection right now.",
                )
        }
    }

    private fun parseEventStream(
        source: okio.BufferedSource,
        bytes: ByteArray,
        progress: ((MirrorProgressEvent) -> Unit)?,
        reflectionChunk: ((MirrorReflectionChunkEvent) -> Unit)?,
    ): MirrorResponsePayload {
        var currentEvent: String? = null
        val dataLines = mutableListOf<String>()

        fun flush(): MirrorResponsePayload? {
            val event = currentEvent ?: run {
                dataLines.clear()
                return null
            }
            val payload = dataLines.joinToString("\n")
            dataLines.clear()
            return when (event) {
                "update" -> {
                    payload.toJsonObjectOrNull()?.let { json ->
                        progress?.invoke(
                            MirrorProgressEvent(
                                stage = json.optString("stage"),
                                message = json.optString("message").takeIf { it.isNotBlank() },
                            ),
                        )
                    }
                    null
                }
                "reflection_chunk" -> {
                    payload.toJsonObjectOrNull()?.let { json ->
                        reflectionChunk?.invoke(
                            MirrorReflectionChunkEvent(
                                chunk = json.optString("chunk"),
                                generatedCharacters = json.optInt("generatedCharacters"),
                            ),
                        )
                    }
                    null
                }
                "reflection" -> reflectionPayloadFromEvent(payload, bytes)
                "error" -> throw MirrorClientError.Server(MirrorErrorCode.Unknown, errorMessageFromSsePayload(payload))
                else -> null
            }
        }

        while (true) {
            val line = source.readUtf8Line() ?: break
            val trimmedLine = line.trim()
            if (trimmedLine.isEmpty()) {
                flush()?.let { return it }
                currentEvent = null
                continue
            }
            when {
                trimmedLine.startsWith("event:") -> {
                    if (currentEvent != null) {
                        flush()?.let { return it }
                    }
                    currentEvent = trimmedLine.removePrefix("event:").trim()
                }
                trimmedLine.startsWith("data:") -> {
                    dataLines += trimmedLine.removePrefix("data:").trim()
                }
            }
        }

        flush()?.let { return it }
        throw MirrorClientError.InvalidResponse
    }
}

internal fun MirrorConfiguration.effectiveBaseUrl(): String {
    val candidate = baseUrl.trim()
    return candidate.ifEmpty { MirrorConfiguration().baseUrl }
}

private fun String.toJsonObjectOrNull(): JSONObject? =
    runCatching { JSONObject(this) }.getOrNull()

private fun JSONObject.toMirrorServerError(): MirrorClientError.Server {
    val error = optJSONObject("error")
    val code = error?.optString("code").toMirrorErrorCode()
    val message = error
        ?.optString("message")
        ?.takeUnless { it.isBlank() }
        ?: "Anky could not return a reflection right now."
    return MirrorClientError.Server(code, message)
}

private fun reflectionPayloadFromEvent(payload: String, bytes: ByteArray): MirrorResponsePayload {
    val json = payload.toJsonObjectOrNull() ?: throw MirrorClientError.InvalidResponse
    val reflection = json.optString("markdown").trim().takeIf { it.isNotEmpty() }
        ?: throw MirrorClientError.InvalidResponse
    val headers = json.optJSONObject("headers")
    val hash = headers.headerValue("X-Anky-Hash") ?: AnkyHasher.sha256Hex(bytes)
    val tags = json.optJSONArray("tags").stringList()
        .ifEmpty { headers.headerValue("X-Anky-Tags").tags().orEmpty() }
    return MirrorResponsePayload(
        hash = hash,
        title = titleFromMarkdown(reflection),
        reflection = reflection,
        tags = tags,
        creditsRemaining = headers.headerValue("X-Anky-Credits-Remaining").creditsRemaining(),
    )
}

private fun errorMessageFromSsePayload(payload: String): String {
    val json = payload.toJsonObjectOrNull() ?: return "Anky could not return a reflection right now."
    val body = json.optJSONObject("body")
    val error = body?.optJSONObject("error")
    return error?.optString("message")?.takeIf { it.isNotBlank() }
        ?: json.optString("message").takeIf { it.isNotBlank() }
        ?: "Anky could not return a reflection right now."
}

private fun String?.toMirrorErrorCode(): MirrorErrorCode =
    when (this) {
        "INVALID_ANKY" -> MirrorErrorCode.InvalidAnky
        "INCOMPLETE_RITUAL" -> MirrorErrorCode.IncompleteRitual
        "MISSING_IDENTITY_VERSION" -> MirrorErrorCode.MissingSignature
        "UNSUPPORTED_IDENTITY_VERSION" -> MirrorErrorCode.InvalidSignature
        "MISSING_ACCOUNT" -> MirrorErrorCode.MissingSignature
        "INVALID_ACCOUNT" -> MirrorErrorCode.InvalidSignature
        "UNSUPPORTED_CHAIN" -> MirrorErrorCode.InvalidSignature
        "INVALID_SIGNATURE_TYPE" -> MirrorErrorCode.InvalidSignature
        "MISSING_SIGNATURE" -> MirrorErrorCode.MissingSignature
        "INVALID_SIGNATURE" -> MirrorErrorCode.InvalidSignature
        "INSUFFICIENT_CREDITS" -> MirrorErrorCode.InsufficientCredits
        "TRIAL_ALREADY_CLAIMED" -> MirrorErrorCode.TrialAlreadyClaimed
        "DUPLICATE_IN_PROGRESS" -> MirrorErrorCode.DuplicateInProgress
        "RATE_LIMITED" -> MirrorErrorCode.RateLimited
        "MIRROR_FAILED" -> MirrorErrorCode.MirrorFailed
        "BODY_TOO_LARGE" -> MirrorErrorCode.BodyTooLarge
        else -> MirrorErrorCode.Unknown
    }

private fun String?.creditsRemaining(): Int? =
    if (this == null || this == "null") null else toIntOrNull()

private fun String?.tags(): List<String>? {
    if (this.isNullOrBlank()) return null
    return runCatching { JSONArray(this).stringList() }.getOrNull()
}

private fun JSONArray?.stringList(): List<String> {
    if (this == null) return emptyList()
    return (0 until length()).mapNotNull { index ->
        optString(index).trim().takeIf { it.isNotEmpty() }
    }
}

private fun JSONObject?.headerValue(name: String): String? {
    if (this == null) return null
    val keys = keys()
    while (keys.hasNext()) {
        val key = keys.next()
        if (key.equals(name, ignoreCase = true)) return optString(key).takeIf { it.isNotBlank() }
    }
    return null
}

private fun okhttp3.Response.isEventStream(): Boolean =
    header("Content-Type")?.contains("text/event-stream", ignoreCase = true) == true

private fun titleFromMarkdown(markdown: String): String {
    val lines = markdown.lineSequence().map { it.trim() }.filter { it.isNotEmpty() }.toList()
    val heading = lines.firstOrNull { it.startsWith("# ") }?.removePrefix("# ")?.trim()
    val fallback = lines.firstOrNull()
    return heading?.takeIf { it.isNotEmpty() }
        ?: fallback?.takeIf { it.isNotEmpty() }
        ?: "reflection"
}
